Improved PathSpec handling for servletName & pathInfo

joakime opened the pull request:
Introduction of MatchedResource and MatchedPath to capture the results of a PathSpec match.

joakime committed:
More work on MatchedResult

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Improved PathSpec usage

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Improved RegexPathSpec signature detection

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Improved RegexPathSpec signature detection with groups

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Fixing typo in PathSpecSet

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Fixing comparison in Invoker

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/MatchedPath.java:
```diff
@@ -0,0 +1,51 @@
+
+package org.eclipse.jetty.http.pathmap;
+
+public interface MatchedPath
```
I think it would be more intuitive to have a `getPathSpec()` method on this class, but not have it on `MatchedResource`

joakime replied:
That would mean MappedResource doesn't really need to exist too, as we can do a simple resource return `<E>` now.

joakime replied:
This change caused too many changes.
Be happy to make it in Jetty 12 though.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -270,6 +284,44 @@ private boolean isWildcardMatch(String path)
+            case ROOT:
+                // Only "/" matches
+                if ("/".equals(path))
+                    return new ServletMatchedPath("", path); // TODO: review this
```
bizarrely I think that is correct!  "" matches "/" and returns "" as the match and "/" as the part after the match !?!?!?!?

joakime replied:
Now using `MatchedPath.from("", path);`

joakime replied:
The logic here is identical to how it was in 9.4.46 too.

See: 

* https://github.com/eclipse/jetty.project/blob/jetty-9.4.46.v20220331/jetty-http/src/test/java/org/eclipse/jetty/http/pathmap/ServletPathSpecTest.java#L122
* https://github.com/eclipse/jetty.project/blob/jetty-9.4.46.v20220331/jetty-http/src/test/java/org/eclipse/jetty/http/pathmap/ServletPathSpecTest.java#L152-L153

https://github.com/eclipse/jetty.project/blob/bc17a0369a11ecf40bb92c839b9ef0a8ac50ea18/jetty-http/src/test/java/org/eclipse/jetty/http/pathmap/ServletPathSpecTest.java#L122

https://github.com/eclipse/jetty.project/blob/bc17a0369a11ecf40bb92c839b9ef0a8ac50ea18/jetty-http/src/test/java/org/eclipse/jetty/http/pathmap/ServletPathSpecTest.java#L152-L153

It surprised me too, and we have many tests elsewhere in jetty that test for ROOT path hits.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -270,6 +284,44 @@ private boolean isWildcardMatch(String path)
+                break;
+            case DEFAULT:
+                // If we reached this point, then everything matches
+                return new ServletMatchedPath(path, null);
```
this should also be a return of a known `MatchedPath`:
```suggestion
                return _matchedPath;
```

gregw replied:
So this will be the main cost of this change (in garbage).  Previously we could just return path here, but now will need to wrap it in a `ServletMatchedPath`.    Probably worth it if we can reduced string comparisons elsewhere.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -291,4 +343,28 @@ public boolean matches(String path)
         }
     }
+
+    public static class ServletMatchedPath implements MatchedPath
```
There is nothing "Servlet" about this.   Just add a static method on `MatchedPath` so you can say `MatchedPath.from(pathMatch, pathInfo)`

joakime replied:
Done.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -291,4 +343,28 @@ public boolean matches(String path)
+
+        public ServletMatchedPath(String servletName, String pathInfo)
+        {
+            this.servletName = servletName;
```
this isn't the servletName, it is the pathMatch!

joakime replied:
Done, using MatchedPath.from() instead.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -182,15 +191,25 @@ public MappedResource<E> getMatch(String path)
 
-            if (mr.getPathSpec().matches(path))
-                return mr;
+            matchedPath = mr.getPathSpec().matched(path);
```
I think we have a big inefficiency here.... even in the existing code.

Consider if we have 3 EXACT mappings, 3 PREFIX and the DEFAULT mappings, then we will:
1. iterate over this loop 7 times.
2. twice we will fire the `group != lastGroup` test and do an efficient Trie lookup
3. once we fire the `group != lastGroup` and do nothing
4. So all 7 iterations come to this line and call `matched(path)`, even though we know for 6 of them that they will not match because they failed in the Trie

We need to change this loop so that if an EXACT or PREFIX match fails in the Trie, then all iterations for the same group are skipped and we don't try the explicit match here for them all.   

This change will more than make up for the small amount of extra garbage

joakime replied:
I think the original loop without the Tries is more efficient (in terms of speed and memory).

I'll work with @lorban and mock up some jmh tests to see what is truely better.

joakime replied:
@gregw  I think we've addressed your concern on this block of code.

joakime committed:
Issue #7891 - Correct no-match NativeWebSocketConfiguration condition

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Introduce MatchedPath.from

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Add toString() impls

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Improving MatchedResource

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Fixing compilation issues

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Better group match optimization logic

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7891 - Better group match optimization logic

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -231,24 +295,48 @@ public boolean put(PathSpec pathSpec, E resource)
                 {
-                    _exactMap = new ArrayTernaryTrie<>((ArrayTernaryTrie<MappedResource<E>>)_exactMap, 1.5);
+                    // This is not a Servlet mapping, turn off optimization on Exact
+                    _optimizedExact = false;
```
I feel that we probably can do optimized exact matches, even for regex etc.   The Trie maps to a MappedResource that includes the PathSpec, so if match produces a non ServletPathSpec match, then we know we have to run it again to ensure we match any groups etc.

But maybe this is OK for now.  Perhaps add a TODO to investigate if an exact map can be done.

joakime replied:
Lets punt on optimized regex for now.

joakime replied:
Added some comments about the concerns on exact/prefix/suffix and non-servlet path specs

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -231,24 +295,48 @@ public boolean put(PathSpec pathSpec, E resource)
-                while (exact != null && !_exactMap.put(exact, entry))
+                if (pathSpec instanceof ServletPathSpec)
+                {
+                    String exact = pathSpec.getDeclaration();
```
I know it is not part of this PR, but can you put a comment in the while below that says what the loop is for (i.e. it is growing the capacity of the trie).  Took me a few minutes to work that out.

joakime replied:
Done.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -231,24 +295,48 @@ public boolean put(PathSpec pathSpec, E resource)
                 {
-                    _prefixMap = new ArrayTernaryTrie<>((ArrayTernaryTrie<MappedResource<E>>)_prefixMap, 1.5);
+                    // This is not a Servlet mapping, turn off optimization on Prefix
+                    _optimizedPrefix = false;
```
Again, we should be able to extract the prefix from any PREFIX_GLOB and put that in the Trie.
So again, for now add a TODO to optimize later.

joakime replied:
It's not quite as easy as you think.
The comments added show what needs to be resolved.
TODO added as well.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -231,24 +295,48 @@ public boolean put(PathSpec pathSpec, E resource)
+                else
+                {
+                    // This is not a Servlet mapping, turn off optimization on Suffix
+                    _optimizedSuffix = false;
```
Hmmm I really don't like all these optimized booleans.  Might just be better to null out the map/trie and have a null mean don't look up.  That is a single test rather than always testing boolean then looking at the map/trie

joakime replied:
That means creating the Trie anew with every .add() or .put()

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -263,21 +351,31 @@ public boolean put(PathSpec pathSpec, E resource)
+                if (exact != null)
+                {
+                    _exactMap.remove(exact);
+                    // TODO: recalculate _optimizeExact
```
Do this before PR can be accepted

joakime replied:
Done.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -231,24 +295,48 @@ public boolean put(PathSpec pathSpec, E resource)
         switch (pathSpec.getGroup())
```
There is an existing bug with this switch.   It should do the `boolean added = _mappings.add(entry);` first and run the switch only if `added==true`

joakime replied:
Fixed in .put() and .remove()

gregw reviewed (changes requested):
A few things from a quick look through.
I don't love that we are not optimizing if there are Regex patterns, but I think that can be a TODO for later.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -270,6 +284,44 @@ private boolean isWildcardMatch(String path)
+        {
+            case EXACT:
+                if (_declaration.equals(path))
+                    return MatchedPath.from(path, null);
```
This should be a final field as it will always be the immutable `MatchedPath.from(_declaration, null)`

joakime replied:
final where?  if you have multiple PathSpecs, you have multiple final fields, which means a lookup/map again.

gregw commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -270,6 +284,44 @@ private boolean isWildcardMatch(String path)
+                    return MatchedPath.from(path, null);
+                break;
+            case PREFIX_GLOB:
+                if (isWildcardMatch(path))
+                {
+                    String pathMatch = path;
+                    String pathInfo = null;
+                    if (path.length() != (_specLength - 2))
+                    {
+                        pathMatch = path.substring(0, _specLength - 2);
+                        pathInfo = path.substring(_specLength - 2);
+                    }
+                    return MatchedPath.from(pathMatch, pathInfo);
```
This also needs to avoid allocation in the common case of "/foo" matching "/foo/*"

```suggestion
                if (isWildcardMatch(path))
                {
                    if (path.length() == (_specLength - 2))
                        return finalFieldAllocatedInConstructor;
                    String pathMatch = anotherFinalFiledAllocatedInConstructor ;
                    String pathInfo = path.substring(_specLength - 2);
                    return MatchedPath.from(pathMatch, pathInfo);
```

joakime replied:
You'll have a map of values, mapping each PathSpec and potential path.

gregw replied:
The pathSpec itself should know it's prefix, so that is what is returned.  No map needed.

gregw commented:
@joakime do you want me to have a go at fixing some of these issues?

joakime committed:
Issue #7891 - Optimized .put() and .remove()

+ Implemented recalculations
+ Added more comments

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented:
I addressed most of the review comments.

The comments about final fields for various things don't make much sense to me, as we are dealing with multiple PathSpec objects in the collection, so having a final field make sense will result in another Map lookup, which kinda defeats the whole final field benefit.

If you can make it work without a new collection/trie/etc, give it a whirl.

gregw commented:
I'll do a PR against this PR addressing some of my performance concerns.... good task for on the train/plane tomorrow.

joakime commented:
@gregw :triangular_flag_on_post: bump :triangular_flag_on_post: : this need attention for the next release.

gregw commented:
Ack

gregw committed:
Introduced final preMatchedPath when possible

gregw commented:
@joakime I pushed a change to this branch that should improve performance.
You'll need to get somebody else to review now.

gregw committed:
Improved match

joakime commented:
+1 from me.
Now lets get @janbartel @lachlan-roberts @sbordet or @lorban to review it (and hopefully approve it)

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/AbstractPathSpec.java:
```diff
@@ -55,7 +60,7 @@ public final boolean equals(Object obj)
     public final int hashCode()
     {
-        return Objects.hash(getDeclaration());
+        return Objects.hash(this.getClass().getSimpleName() + ":" + getDeclaration());
```
Implementation of `hashCode()` should be based on the same items used for `equals()`.
Consider:

```suggestion
        return Objects.hash(getGroup().ordinal(), getSpecLength(), getDeclaration(), getClass().getName());
```

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/MatchedPath.java:
```diff
@@ -0,0 +1,81 @@
+        @Override
+        public String toString()
+        {
+            return "MatchedPath.EMPTY";
```
Better to use `getClass().getSimpleName()`.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/MatchedPath.java:
```diff
@@ -0,0 +1,81 @@
+            @Override
+            public String toString()
+            {
+                return "MatchedPath.from[pathMatch=" + pathMatch + ", pathInfo=" + pathInfo + "]";
```
Better to use `getClass().getSimpleName()`.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/MatchedResource.java:
```diff
@@ -0,0 +1,70 @@
+
+import java.util.Map;
+
+public class MatchedResource<E>
```
Javadocs? Why it is generic on `E` and what types `E` can be?

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/MatchedPath.java:
```diff
@@ -0,0 +1,81 @@
+
+package org.eclipse.jetty.http.pathmap;
+
+public interface MatchedPath
```
Javadocs?

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathMappings.java:
```diff
@@ -292,8 +378,50 @@ public boolean remove(PathSpec pathSpec)
+                        // Recalculate _optimizeExact
+                        _optimizedExact = _mappings.stream()
+                            .filter((mapping) -> mapping.getPathSpec().getGroup() == PathSpecGroup.EXACT)
+                            .allMatch((mapping) -> mapping.getPathSpec() instanceof ServletPathSpec);
```
Consider moving this logic into a method and reused it in the other `case` blocks.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/PathSpec.java:
```diff
@@ -25,6 +25,17 @@
+    static PathSpec from(String pathSpecString)
+    {
+        if (pathSpecString == null)
+            throw new RuntimeException("Path Spec String must start with '^', '/', or '*.': got [" + pathSpecString + "]");
```
Should not this be a NPE? 
Furthermore, if it's non-null, there is no check that `pathSpecString` starts with ^, /, etc. The only check below is for ^ to distinguish between regex or not.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -181,16 +264,88 @@ public Pattern getPattern()
+        return null;
+    }
+
+    public class RegexMatchedPath implements MatchedPath
```
Can be private?

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -181,16 +264,88 @@ public Pattern getPattern()
+        @Override
+        public String toString()
+        {
+            return "RegexMatchedPath[" +
```
Use class.simpleName.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/ServletPathSpec.java:
```diff
@@ -90,13 +96,15 @@ public ServletPathSpec(String servletPathSpec)
             group = PathSpecGroup.PREFIX_GLOB;
             prefix = servletPathSpec.substring(0, specLength - 2);
             suffix = null;
+            _preMatchedPath = MatchedPath.from(prefix, null);
```
Store in a local variable and assign below.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/UriTemplatePathSpec.java:
```diff
@@ -421,4 +440,62 @@ public String[] getVariables()
         return _variables;
     }
+
+    public static class UriTemplateMatchedPath implements MatchedPath
```
Make private.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/UriTemplatePathSpec.java:
```diff
@@ -421,4 +440,62 @@ public String[] getVariables()
+        @Override
+        public String toString()
+        {
+            return "UriTemplateMatchedPath[" +
```
Use class.simpleName.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/RegexPathSpec.java:
```diff
@@ -38,34 +59,80 @@ public RegexPathSpec(String regex)
-                    if (!inGrouping)
+                    if (!inTextList && !inQuantifier)
                         pathDepth++;
                     break;
```
Should be there a `case` for `<` and `>` (for named groups)?

joakime replied:
Not needed, the group detects a signature `g` properly and that sets up the signature check later approriately.
Other uses of `<` and `>` are allowed.
New test case added.

sbordet commented on jetty-http/src/test/java/org/eclipse/jetty/http/pathmap/RegexPathSpecTest.java:
```diff
@@ -151,6 +152,42 @@ public void testSuffixSpec()
+        assertNotMatches(spec, "/a.do");
+        assertNotMatches(spec, "/a/middle");
+        assertNotMatches(spec, "/middle");
+    }
```
Is there a test for named groups?

joakime replied:
Added testcase

joakime committed:
Changes from review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>
